### PR TITLE
Implement a max jobs per worker budget

### PR DESCRIPTION
**Pull request by staabm**
implements https://github.com/rectorphp/rector/issues/8192

### CODE REVIEW

File: packages/Parallel/Application/ParallelFileProcessor.php [hunk 1]
```diff
@@ -42,6 +42,10 @@ final class ParallelFileProcessor
      * @var int
      */
     private const SYSTEM_ERROR_LIMIT = 50;
+    /**
+     * @var int
+     */
+    private const MAX_JOBS_PER_WORKER = 10;
```
Comment (staabm) [inline packages/Parallel/Application/ParallelFileProcessor.php]: looking at the RAM consumption I think 10 is a good value. I was wondering why the parallel job size is with 20 pretty big in contrast.

we could play with those numbers of course.

adjusting this to 10 I could run rector across mautic with

```
    $rectorConfig->paths([
        __DIR__.'/app/',
        __DIR__.'/plugins',
    ]);
```

with peaks of 16 GB RAM

**Comment by staabm** [comment]
will be AFK for a few hours

File: packages/Parallel/Application/ParallelFileProcessor.php [hunk 2]
```diff
@@ -195,12 +214,20 @@ function (array $json) use (
...
+                        ($processSpawner)();
+                        return;
+                    }
                     if ($jobs === []) {
                         $this->processPool->quitProcess($processIdentifier);
                         return;
                     }
 
                     $job = array_pop($jobs);
+                    $fileBudgetPerProcess[$processIdentifier]--;
```
Comment (staabm) [inline packages/Parallel/Application/ParallelFileProcessor.php]: Maybe I should move this one line below

Comment (samsonasik) [reply packages/Parallel/Application/ParallelFileProcessor.php]: You should use `--$fileBudgetPerProcess[$processIdentifier]`

```suggestion
                    --$fileBudgetPerProcess[$processIdentifier];
```

File: packages/Parallel/Application/ParallelFileProcessor.php (hunk 1 above)
Comment (samsonasik) [reply packages/Parallel/Application/ParallelFileProcessor.php]: could you comment what is the different between `MAX_JOBS_PER_WORKER` vs `$jobSize` for future reference? 

It defined at as 3rd arg and param at:

https://github.com/rectorphp/rector-src/blob/9f29bf8434e78a86efe97a08ef3e84c8db2f4a94/config/config.php#L16-L16

and

https://github.com/rectorphp/rector-src/blob/9f29bf8434e78a86efe97a08ef3e84c8db2f4a94/packages/Config/RectorConfig.php#L62

File: packages/Config/RectorConfig.php [hunk 3]
```diff
@@ -59,7 +59,7 @@ public function disableParallel(): void
         SimpleParameterProvider::setParameter(Option::PARALLEL, false);
     }
 
-    public function parallel(int $seconds = 120, int $maxNumberOfProcess = 16, int $jobSize = 20): void
+    public function parallel(int $seconds = 120, int $maxNumberOfProcess = 16, int $jobSize = 15): void
```
Comment (staabm) [inline packages/Config/RectorConfig.php:62]: reduced the jobsize, which leads to less memory used

File: packages/Parallel/Application/ParallelFileProcessor.php [hunk 4]
```diff
@@ -195,16 +216,24 @@ function (array $json) use (
...
+                        ($processSpawner)();
+                        return;
+                    }
                     if ($jobs === []) {
                         $this->processPool->quitProcess($processIdentifier);
                         return;
                     }
 
-                    $job = array_pop($jobs);
+                    $jobsChunk = array_pop($jobs);
```
Comment (staabm) [inline packages/Parallel/Application/ParallelFileProcessor.php:232]: renaming because the previous name suggested it would be a single job

**Comment by staabm** [comment]
with the parameter adjustments at this point we can run rector across mautic with

```
    $rectorConfig->paths([
        __DIR__.'/app/',
        __DIR__.'/plugins',
    ]);
```

with a peak of 14.75 GB. it never grows higher then that:
```
time php ../rector-src/bin/rector.php --dry-run --no-progress-bar --no-diffs --clear-cache
                                               
 [OK] 10 files would have changed (dry-run) by Rector                                                                   
                                                                                                                       
php ../rector-src/bin/rector.php --dry-run --no-progress-bar --no-diffs   379.85s user 11.00s system 798% cpu 48.939 total
```

----

with rector 0.18.2 it peaks at 16.5 GB RAM and takes

```
➜  mautic git:(5.x) ✗ time php ../rector-src/bin/rector.php --dry-run --no-progress-bar --no-diffs --clear-cache
                                                                                                                        
 [OK] 30 files would have changed (dry-run) by Rector                                                                                                                                                                                      

php ../rector-src/bin/rector.php --dry-run --no-progress-bar --no-diffs   516.54s user 8.36s system 766% cpu 1:08.51 total
```

=> with this PR (and all the already merged optimizations) we are 20 seconds faster and take ~1.75 GB less memory

File: packages/Config/RectorConfig.php (hunk 3 above)
Comment (samsonasik) [reply packages/Config/RectorConfig.php:62]: `config/config.php` need to be updated as well:

https://github.com/rectorphp/rector-src/blob/038f8092eaeeaaf044dbd5685d89bc7636143bbc/config/config.php#L16

could you also update the different between $jobSize vs `MAX_CHUNKS_PER_WORKER` ? Thank you.

Comment (staabm) [reply packages/Config/RectorConfig.php:62]: I dropped the args in this file, as these just reflect the defaults and therefore don't need to be kept in sync

> could you also update the different between $jobSize vs MAX_CHUNKS_PER_WORKER ?

where?

File: packages/Parallel/Application/ParallelFileProcessor.php [hunk 5]
```diff
@@ -42,6 +42,12 @@ final class ParallelFileProcessor
      * @var int
      */
     private const SYSTEM_ERROR_LIMIT = 50;
+    /**
+     * the number of chunks a worker can process before getting killed
+     *
+     * @var int
+     */
+    private const MAX_CHUNKS_PER_WORKER = 8;
```
Comment (samsonasik) [inline packages/Parallel/Application/ParallelFileProcessor.php:51]: it is a bit ambiguous the different between `$jobSize` vs `MAX_CHUNKS_PER_WORKER`, could you write a comment here? Thank you.

File: packages/Config/RectorConfig.php (hunk 3 above)
Comment (samsonasik) [reply packages/Config/RectorConfig.php:62]: here seems ok

↓

https://github.com/rectorphp/rector-src/pull/4965#pullrequestreview-1618817032

**Review by samsonasik — APPROVED** [review]
@TomasVotruba let's quick merge and revert in case of it doesn't work :)

Thank you @staabm

**Comment by samsonasik** [comment]
@staabm I am testing in CodeIgniter 4 project, when we define:

```php
$rectorConfig->parallel(240, 8, 1);
```

the procss got 2 times slower, but when we define default value:

```php
$rectorConfig->parallel();
```

the performance seems equals, I mention you in PR test at:

- https://github.com/codeigniter4/CodeIgniter4/pull/7918

so in case other CodeIgniter4 project maintainer has different use case, they can give you feedback faster :)

**Comment by staabm** [comment]
> the procss got 2 times slower,

thats expected. the smaller the jobsize is and the smaller the chunks-per-worker is, the slower the rector job will get.
but it also uses less memory.

the higher the jobsize and the higher the chunks-per-worker are, rector gets faster but requires more memory.

its a balance between memory and speed. before this PR all default settings were set for maximum speed but without limits for memory. thats the actual problem the bigger the project gets we analyze.

with this PR we have a mean to cut the memory peaks.

performance and memory also depends whether rector just scans the codebase and does not need todo anything or whether it needs to refactor something

**Comment by staabm** [comment]
@Flyingmana @yguedidi you might be interessted in testing rector@dev-main which now uses process re-spawning to recover from memory leaks and memory spikes

File: packages/Config/RectorConfig.php (hunk 3 above)
Comment (TomasVotruba) [reply packages/Config/RectorConfig.php:62]: Looks good :+1:

**Pull request merged**